### DISCUSSION TITLE
[Synology DSM] Any command outputs "brew.sh: line 604: /etc/os-release: No such file or directory"

### DISCUSSION THREAD

**Discussion by ogerardin** [Homebrew/discussions#4670] [Everyday usage] [answered]
### Output of `brew config`

```shell
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 604: /etc/os-release: No such file or directory
HOMEBREW_VERSION: 4.1.1-17-gd0f7d2b
ORIGIN: https://github.com/Homebrew/brew
HEAD: d0f7d2b01173d3f01107329af3064c9c9a5ab5c5
Last commit: 88 minutes ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 61239c4c7192787f360874ea7d7d8aa3361ae2c1
Core tap last commit: 5 months ago
Core tap branch: master
Core tap JSON: 25 Jul 09:05 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 1
HOMEBREW_TEMP: /var/services/homes/olivier/tmp
Homebrew Ruby: 2.6.10 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: single-core 64-bit unknown_0x6_0x5c
Clang: N/A
Git: 2.41.0 => /home/linuxbrew/.linuxbrew/bin/git
Curl: 7.79.1 => /bin/curl
Kernel: Linux 4.4.180+ x86_64 GNU/Linux
OS: Unknown
Host glibc: 2.26
/usr/bin/gcc: 12.3.0
/usr/bin/ruby: N/A
glibc: 2.35_1
gcc@11: 11.4.0
gcc: 13.1.0
xorg: N/A
```

### Output of `brew doctor`

```shell
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 604: /etc/os-release: No such file or directory
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  gcc@5
  isl@0.18
```

### Description of issue

Since a recent update, any brew command outputs:
`/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 604: /etc/os-release: No such file or directory`
before anything else.
It doesn't seem to affect founctionality though.

The system is a Synology NAS with DSM 7.

**Comment by SMillerDev** [comment]
Do you have an `lsb_release` command? And if so, what does it output?

**Reply by ogerardin** [reply]
No `lsb_release` command...

**Comment by osalbahr** [comment]
What’s the output of `brew install --verbose --debug <formula>`? (where \<formula\> is any formula you don’t have installed). I think the error message you see would still be at the very beginning, just double-checking.

**Reply by osalbahr** [reply]
Actually, what’s the output of `file /etc/os-release && cat /etc/os-release`? It looks like the script assumes that `/etc/os-release` exists on all Linux systems, but if that’s not the case then one more check is needed in `brew.sh`.

Side note: I agree that this probably doesn’t affect functionality.

**Reply by ogerardin** [reply]
Obviously `/etc/os-release` does not exist on this system
```
$ file /etc/os-release && cat /etc/os-release
/etc/os-release: cannot open `/etc/os-release' (No such file or directory)
cat: /etc/os-release: No such file or directory
```
No sure what this file is used for when it's there, but at least the message should be redirected to `/dev/null`

**Reply by SMillerDev** [reply]
Great, so you seem to be on the only Linux distro without that file. I guess we'll have to find another way to detect what distro that is.

**Reply by ogerardin** [reply]
As stated it's Synology DSM 7, and since the absence of this file doesn't seem to affect functionality, it means brew doesn't rely on it... 
Generally speaking I believe it's smarter to tests for features than to determine the distro and infer the features from that.

Synology has a `/etc/VERSION` file but it does not match the info contained in os-release:
```
$ cat /etc/VERSION
majorversion="7"
minorversion="1"
major="7"
minor="1"
micro="1"
productversion="7.1.1"
buildphase="GM"
buildnumber="42962"
smallfixnumber="6"
nano="6"
base="42962"
builddate="2023/05/29"
buildtime="20:01:57"
```
There are ways to determine if it's a Synology distro but I don't think there's a point as it would be non-standard.

**Reply by SMillerDev** [reply]
> As stated it's Synology DSM 7, and since the absence of this file doesn't seem to affect functionality, it means brew doesn't rely on it...

It is used in `brew doctor` and in analytics for maintainers to figure out what OS you're using and how to support it. I'd say brew does rely on it.

**Reply by osalbahr** [reply]
Homebrew does not rely on it, yeah (for basic functionality, that is). However, in the `brew config` output:

    OS: Unknown

which might make it more difficult to debug issues if they were to arrise.

I’m curious, what do you get when running `fastfetch` or [`neofetch` (the one by hyfetch)](https://github.com/hykilpikonna/hyfetch/blob/master/neofetch)?

**Reply by osalbahr** [reply]
> but at least the message should be redirected to /dev/null

I see the argument. And you might be interested in doing that locally for the meantime. I don’t think it’s a good idea to simply mask the error message for everyone. Or, if you’re interested, you can PR conditionally handling your case in `brew.sh:HOMEBREW_OS_VERSION`. Maybe something similar to what `neofetch` does. I think that would do it.

**Reply by ogerardin** [reply]
I don't think the goal is be as complete as neofetch so it makes no sense to implement a full detection logic. For 99% of users the current detection via `os-release` will work, and for the handful of us that are using a distro without `os-release`, it's OK to say "OS: Unknown".

However, I still believe this message should be hidden in normal use because it does not affect functionality at all, it's just for diagnosis purposes. As an end user I don't see why I would need to be made aware of that if everything still works fine. It makes sense though when invoking `brew doctor` or `brew config` which are specifically for diagnosis purposes.

**Comment by ogerardin** [comment]
> I’m curious, what do you get when running `fastfetch` or [`neofetch` (the one by hyfetch)](https://github.com/hykilpikonna/hyfetch/blob/master/neofetch)?

`fastfetch` has no clue (btw /volume2 is NOT removable):

```
        #####            [redacted]
       #######           -----------------
       ##O#O##           Host: DS918+ (1.0)
       #######           Kernel: 4.4.180+
     ###########         Uptime: 41 days, 19 hours, 43 mins
    #############        Packages: 69 (brew), 80 (pkgtool)
   ###############       Shell: sh 4.4.23
   ################      Terminal: /dev/pts/0
  #################      CPU: Intel(R) Celeron(R) J3455 (4) @ 1.501 GHz
#####################    Memory: 2.90 GiB / 11.55 GiB (25%)
#####################    Disk (/): 1.85 GiB / 2.28 GiB (81%)
  #################      Disk (/volume2): 8 TiB / 10 TiB (75%) [Removable]
                         Locale: en_US.utf8
```

`neofetch` does a better job and recognize Synology! Yay!
```
        #####           [redacted]
       #######          -----------------
       ##O#O##          OS: Synology DSM 7.1.1-42962 Update 6 x86_64
       #######          Host: DS918+ 1.0
     ###########        Kernel: 4.4.180+
    #############       Uptime: 41 days, 19 hours, 46 mins
   ###############      Packages: 69 (brew)
   ################     Shell: sh
  #################     Terminal: /dev/pts/0
#####################   CPU: Intel Celeron J3455 (4) @ 1.5GHz
#####################   Memory: 3.29 GiB / 11.54 GiB (28%)
  #################     Network: 1 Gbps
                        BIOS: INSYDE Corp. (11/11/2019)
```

**Comment by ogerardin** [comment]
And I was not wrong, it tests `/etc/VERSION` :)
```
elif [[ -f /etc/VERSION ]]; then
                # Source the DSM VERSION file
                source /etc/VERSION
                distro="Synology DSM ${productversion}-${buildnumber} Update ${smallfixnumber}"
```

**Reply by osalbahr** [reply]
Does it immediately assume that it’s Synology just because of the existence of `/etc/VERSION`?

**Reply by ogerardin** [reply]
Apparently yes...
https://github.com/hykilpikonna/hyfetch/blob/0385fd836d55af63464be6e2fd87f17a5a17d23f/neofetch#L1238

**Reply by osalbahr** [reply]
Yeah I think you’re right. This is when it was added:
- https://github.com/dylanaraps/neofetch/pull/1666

**Comment by vsenn** [comment]
This workaround to eliminate Homebrew warning works just fine for me on DSM 7.1

`sudo install -m 755 /dev/stdin /etc/os-release <<EOF
#!/bin/bash
echo "PRETTY_NAME=\"\$(source /etc.defaults/VERSION && echo \${os_name} \${productversion}-\${buildnumber} Update \${smallfixnumber})\""
EOF`

No more `/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.sh: line 604: /etc/os-release: No such file or directory` after this quick fix

**Reply by ogerardin** [reply, marked as answer]
Nice one! 
I had to split it on several lines otherwise the heredoc doesn't work.
```
sudo install -m 755 /dev/stdin /etc/os-release <<EOF 
#!/bin/bash 
echo "PRETTY_NAME=\"\$(source /etc.defaults/VERSION && echo \${os_name} \${productversion}-\${buildnumber} Update \${smallfixnumber})\"" 
EOF

```

**Reply by vsenn** [reply]
Yes, in your formatting. Don't know, how to add a carriage return 😄 Thanks a ton for entire brew\DSM hack!

**Reply by ogerardin** [reply]
Sorry I had to mark my own reply as answer because of the formatting...
FYI to retain CRs you need to use a [block quote](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) (triple backtick)
